File: plyushkin-service/src/main/java/com/plyushkin/budget/expense/ExpenseNoteCategoryId.java
package com.plyushkin.budget.expense;

import static lombok.AccessLevel.PROTECTED;

import jakarta.persistence.Column;
import jakarta.persistence.Embeddable;
import java.io.Serial;
import java.io.Serializable;
import lombok.EqualsAndHashCode;
import lombok.NoArgsConstructor;

@Embeddable
@NoArgsConstructor(access = PROTECTED)
@EqualsAndHashCode
public class ExpenseNoteCategoryId implements Serializable {
  @Serial
  private static final long serialVersionUID = 1L;

  @Column(name = "id", updatable = false)
  private long value;

  public static ExpenseNoteCategoryId create(long value) throws InvalidExpenseNoteIdException {
    if (value <= 0) {
      throw new InvalidExpenseNoteIdException(
          "Value should be positive but it is: " + value
      );
    }
    ExpenseNoteCategoryId expenseNoteCategoryId = new ExpenseNoteCategoryId();
    expenseNoteCategoryId.value = value;
    return expenseNoteCategoryId;
  }

  public static class InvalidExpenseNoteIdException extends Exception {

    public InvalidExpenseNoteIdException(String message) {
      super(message);
    }
  }
}

File: plyushkin-backend/src/main/kotlin/com/plyushkin/budget/expense/ExpenseNoteCategoryId.kt
package com.plyushkin.budget.expense

import arrow.core.Either

data class ExpenseNoteCategoryId internal constructor(val value: Long) {
    companion object {}
}

class InvalidExpenseNoteCategoryId(message: String) : IllegalArgumentException(message)

fun ExpenseNoteCategoryId.Companion.create(value: Long): Either<InvalidExpenseNoteCategoryId, ExpenseNoteCategoryId> =
    value.let {
        if (it <= 0) {
            Either.Left(InvalidExpenseNoteCategoryId("Value should be positive: $value"))
        } else {
            Either.Right(ExpenseNoteCategoryId(value))
        }
    }
